libtailscale/backend.go

```diff
@@ -346,3 +332,29 @@ func (a *App) newBackend(dataDir, directFileRoot string, appCtx AppContext, stor
 	}()
 	return b, nil
 }
+
+func (b *backend) isConfigNonNilAndDifferent(rcfg *router.Config, dcfg *dns.OSConfig) bool {
+	if reflect.DeepEqual(rcfg, b.lastCfg) && reflect.DeepEqual(dcfg, b.lastDNSCfg) {
+		b.logger.Logf("isConfigNonNilAndDifferent: no change to Routes or DNS, ignore")
+		return false
+	}
+	return rcfg != nil
+}
+
+func (a *App) closeVpnService(err error, b *backend) {
+	log.Printf("VPN update failed: %v", err)
+
+	mp := new(ipn.MaskedPrefs)
+	mp.WantRunning = false
+	mp.WantRunningSet = true
+
+	_, localApiErr := a.EditPrefs(*mp)
+	if err != nil {
+		log.Printf("localapi edit prefs error %v", localApiErr)
+	}
+
+	b.lastCfg = nil
+	b.CloseTUNs()
+
+	vpnService.service = nil
```

Comment:
I didn't use vpnService.service.Close() because this is a call from go to Android, and IPNService.close() is mostly calling go functions. https://github.com/tailscale/tailscale-android/pull/522 moves the updating VPN status call out of the close function into the service APIs that get called when the VPN is closed. closeVpnService contains the same logic that vpnService.service.Close() does. I just noticed that the other thing that vpnService.service.Close() does that closeVpnService() doesn't do is it calls stopForeground(true), which removes the service from the foreground state and allows it to be killed if there are memory constraints. I think this is actually not what we want; I think we want to call stopSelf(), since we want the VPN service to be killed regardless of memory needs.

Reply:
I think this should work.

One other thing that I just noticed and that we might want to address (but not necessarily in this PR) is that `netns.SetAndroidProtectFunc(nil)` might never be called. This is orthogonal to the changes in this PR and has probably been in this state for a while. I'm not sure if it causes any real issues, but it's still something to keep in mind.